src/lib.rs

```diff
@@ -168,32 +487,124 @@ impl Child {
 pub struct ExitStatus(u32);
 
 impl ExitStatus {
+    /// Success is defined as a zero exit status.
+    ///
+    /// This function return `true` if the `ExitStatus` is zero and `false`
+    /// otherwise.
     pub fn success(&self) -> bool {
         self.0 == 0
     }
 
+    /// Returns the exit code of the process
     pub fn code(&self) -> u32 {
         self.0
     }
 }
 
-unsafe fn close_handles(process_info: &PROCESS_INFORMATION) {
-    CloseHandle(process_info.hProcess);
-    CloseHandle(process_info.hThread);
+use thiserror::Error;
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// Error type of a failed function on a child process.
+///
+/// Note that this error type is not the same of a non-zero exit status. The
+/// error code linked to the error type is the result of the
+/// [`GetLastError`][get-last-error] function. The variants give some context
+/// to the user but if you want more information about an error code, you can
+/// look at the [`System Error Codes`][system-error-codes].
+///
+/// [get-last-error]: https://docs.microsoft.com/en-us/windows/win32/api/errhandlingapi/nf-errhandlingapi-getlasterror
+/// [system-error-codes]: https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes
+#[derive(Error, Debug)]
+pub enum Error {
+    /// An error occurred when calling [`CreateProcessW`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw)
+    #[error("cannot create process (code {:#x})", 0)]
+    CreationFailed(u32),
+
+    /// An error occurred when calling [`WaitForSingleObject`](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject).
+    #[error("cannot wait process (code {:#x})", 0)]
+    WaitFailed(u32),
+
+    /// An error occurred when calling [`TerminateProcess`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess)
+    #[error("cannot kill process (code {:#x})", 0)]
+    KillFailed(u32),
+
+    /// An error occurred when calling [`GetExitCodeProcess`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getexitcodeprocess)
+    #[error("cannot get exit status (code {:#x})", 0)]
+    GetExitCodeFailed(u32),
+
+    /// An error occurred when calling
+    /// [`GetProcessId`](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getprocessid).
+    #[error("cannot get process id (code {:#x})", 0)]
+    GetProcessIdFailed(u32),
+}
+
+impl Error {
+    /// Return the system error code of the Error
+    ///
+    /// This error code isn't formatted like the code in the error string.
+    ///
+    /// To more information about this [codes][system-error-codes.]
+    ///
+    /// [system-error-codes]: https://docs.microsoft.com/en-us/windows/win32/debug/system-error-codes
+    pub fn code(&self) -> u32 {
+        match *self {
```

Comment:
It looks like your type inherited Copy for some reason... :thinking: this is weird...

Reply:
I just checked by compiling the doc but it doesn't show Copy... :thinking: probably some match-pattern magic.

This is fine them.